delocate/delocating.py

```diff
@@ -80,29 +124,79 @@ def delocate_tree_libs(lib_dict, lib_path, root_path):
             if not exists(required):
                 raise DelocationError('library "{0}" does not exist'.format(
                     required))
-            copied_libs[required] = requirings
+            # Copy requirings to preserve it since it will be modified later.
+            needs_copying[required] = dict(requirings)
             copied_basenames.add(r_ed_base)
         else:  # Is local, plan to set relative loader_path
-            delocated_libs.add(required)
-    # Modify in place now that we've checked for errors
-    for required in copied_libs:
-        shutil.copy(required, lib_path)
-        # Set rpath and install names for this copied library
-        for requiring, orig_install_name in lib_dict[required].items():
-            req_rel = relpath(rp_lib_path, dirname(requiring))
-            set_install_name(requiring, orig_install_name,
-                             '@loader_path/{0}/{1}'.format(
-                                 req_rel, basename(required)))
-    for required in delocated_libs:
+            needs_delocating.add(required)
+    return needs_copying, needs_delocating
+
+
+def _copy_required_libs(
+    lib_dict,  # type: Mapping[Text, Mapping[Text, Text]]
+    lib_path,  # type: Text
+    root_path,  # type: Text
+    libraries_to_copy  # type: Iterable[Text]
+):
+    # type (...) -> Tuple[Dict[Text, Dict[Text, Text]], Set[Text]]
+    """ Copy libraries outside of root_path to lib_path.
+
+    Returns
+    -------
+    updated_lib_dict : dict
+        A copy of `lib_dict` modified so that dependencies now point to
+        the copied library destinations.
+    needs_delocating : set of str
+        A set of the destination files, these need to be delocated.
+    """
+    lib_dict = _copy_lib_dict(lib_dict)
```

Comment:
I prefer the casting solution by a fair margin - as in:

```
lib_dict_out = cast(
        Dict[Text, Mapping[Text, Text]],
        deepcopy(lib_dict))
```

Why?   Because it makes it clear _why_ we need to do this - which is a) to avoid stamping on the input dictionary (the `deepcopy`) and b) to satisfy the type-system (`cast`).   At the moment, I think someone reading the code will think what I thought, which is "Hmm, that's funny, I wonder why they are doing that".  Of course we could answer it with comments, but the `cast` code above answers it immediately, without need for a new function to look at.

Reply:
I keep pushing back on this idea because `cast` isn't doing what you think it's doing.  In this case a bad cast is harmless because it's a private function and we know for sure the actual type being passed in, but if a Mapping type which isn't a Dict is ever passed in then this cast will break code depending on the types being dict.  The methods expected in Python's dict type such as `dict.update` would be missing for example.

Also note that `lib_dict_out` has to be `Dict[Text, Dict[Text, Text]]`.  I mentioned `Dict[Text, Mapping[Text, Text]]` as an undesirable type I was getting before when Mypy didn't recognize what I was trying to do.

If I change the input type of `lib_dict` to be `Dict[Text, Dict[Text, Text]]` then `deepcopy` will work how you expect it to without needing any kind of cast or trick.  This would be okay to do on private function.  Any comments or name changes can also be added to make it more clear, but I wanted to get rid of the original variable so that it couldn't be tampered with accidentally.

Reply:
I did know what cast was doing, but it's true I hadn't considered non-dict types as input values.   I still think it would be more readable if the copy has a different name from the original - and some comments here would be good, so that no-one else finds themselves in the same dark corner as me ...  On the lines of:

```
# Make copy of `lib_dict` where mapping and mapping values are dict copies.
# This prevents us overwriting the input `lib_dict` and enforces mutable dicts
# in the working variable and output.
```